src/network/network_gui.cpp

```diff
@@ -1761,166 +1743,534 @@ static void PopupClientList(ClientID client_id, int x, int y)
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_SERVER_NAME_QUERY_CAPTION, NETWORK_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
 
-		uint width = 100; // Default width
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			width = std::max(width, GetStringBoundingBox(ci->client_name).width);
+			case WID_CL_CLIENT_NAME_EDIT:
+				this->query_widget = WID_CL_CLIENT_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.client_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_QUERY_CAPTION, NETWORK_CLIENT_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
+
+			case WID_CL_SERVER_VISIBILITY:
+				if (!_network_server) break;
+
+				ShowDropDownMenu(this, _server_visibility_dropdown, _settings_client.network.server_advertise, WID_CL_SERVER_VISIBILITY, 0, 0);
+				break;
+
+			case WID_CL_MATRIX: {
+				ButtonCommon *button = this->GetButtonAtPoint(pt);
+				if (button == nullptr) break;
+
+				button->OnClick(this, pt);
+			} break;
+		}
+	}
+
+	bool OnTooltip(Point pt, int widget, TooltipCloseCondition close_cond) override
+	{
+		switch (widget) {
+			case WID_CL_SERVER_NAME:
+				GuiShowTooltips(this, STR_NETWORK_CLIENT_LIST_SERVER_NAME_TOOLTIP, 0, nullptr, close_cond);
+				return true;
+
+			case WID_CL_CLIENT_NAME:
+				GuiShowTooltips(this, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_TOOLTIP, 0, nullptr, close_cond);
+				return true;
+
+			case WID_CL_MATRIX: {
+				ButtonCommon *button = this->GetButtonAtPoint(pt);
+				if (button == nullptr) return false;
+
+				GuiShowTooltips(this, button->tooltip, 0, nullptr, close_cond);
+				return true;
+			};
 		}
 
-		size->width = WD_FRAMERECT_LEFT + this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT + width + WD_FRAMERECT_RIGHT;
+		return false;
 	}
 
-	void OnPaint() override
+	void OnDropdownSelect(int widget, int index) override
 	{
-		/* Check if we need to reset the height */
-		if (!this->CheckClientListHeight()) return;
+		switch (widget) {
+			case WID_CL_SERVER_VISIBILITY:
+				if (!_network_server) break;
 
-		this->DrawWidgets();
+				_settings_client.network.server_advertise = (index != 0);
+				break;
+
+			default:
+				NOT_REACHED();
+		}
+
+		this->SetDirty();
 	}
 
-	void DrawWidget(const Rect &r, int widget) const override
+	void OnQueryTextFinished(char *str) override
+	{
+		if (str == nullptr) return;
+
+		switch (this->query_widget) {
+			default: NOT_REACHED();
+
+			case WID_CL_SERVER_NAME_EDIT: {
+				if (!_network_server) break;
+
+				uint index;
+				GetSettingFromName("network.server_name", &index);
+				SetSettingValue(index, StrEmpty(str) ? "My Server" : str);
+				this->InvalidateData();
+				break;
+			}
+
+			case WID_CL_CLIENT_NAME_EDIT: {
+				uint index;
+				GetSettingFromName("network.client_name", &index);
+				SetSettingValue(index, StrEmpty(str) ? "Player" : str);
```

Comment:
Counting the change in NetworkInitialize this is the fourth place where "Player" is hardcoded. Practically we just want a function that validates the name and sets it to Player when required. However, I would consider that out-of-scope for this. I got some ideas for this, though I have no idea yet how much impact that is going to have.

Reply:
I wanted to suggest I can cook something up for this PR, but if you have ideas, we leave it out for now :) This part needs some cleaning anyway, as it is rather weird how it is currently done :)

Also, people really do not like that you can have an empty name (or be named "Player"). I have been considering making it mandatory for you to set a name. But that too is out of scope of this PR :D